compiler/sem/pragmas.nim

```diff
@@ -1657,65 +1418,324 @@ proc prepareSinglePragma(
+  of wHint:
+    let (s, err) = strLitToStrOrErr(c, it)
+    result =
+      if err.isNil:
+        recordPragma(c, it, "hint", s)
+        c.config.localReport(it.info, reportStr(rsemUserHint, s))
+        it
+      else:
+        err
+  of wWarning:
+    let (s, err) = strLitToStrOrErr(c, it)
+    result =
+      if err.isNil:
+        recordPragma(c, it, "warning", s)
+        c.config.localReport(it.info, reportStr(rsemUserWarning, s))
+        it
+      else:
+        err
+  of wError:
+      result = it
+      let s = getStrLitNode(c, it)
+      case s.kind:
+        of nkError:
+          result = s # err
+        else:
+          recordPragma(c, it, "error", s.strVal)
+          result = c.config.newError(
+            it, PAstDiag(kind: adSemCustomUserError, errmsg: s.strVal))
+  of wFatal:
+    result = c.config.newError(it, PAstDiag(kind: adSemFatalError))
+  of wDefine:
+    result = processDefine(c, it)
+  of wUndef:
+    result = processUndef(c, it)
+  of wCompile:
+    result = processCompile(c, it)
+  of wLink:
+    result = processLink(c, it)
+    result = it
+  of wPassl:
+    let (s, err) = strLitToStrOrErr(c, it)
+    result =
+      if err.isNil:
+        extccomp.addLinkOption(c.config, s)
+        recordPragma(c, it, "passl", s)
+        it
+      else:
+        err
+  of wPassc:
+    let (s, err) = strLitToStrOrErr(c, it)
+    result =
+      if err.isNil:
+        extccomp.addCompileOption(c.config, s)
+        recordPragma(c, it, "passc", s)
+        it
       else:
-        result = invalidPragma(c, it)
+        err
+  of wLocalPassc:
+    assert owner.kind == skModule
+    let (s, err) = strLitToStrOrErr(c, it)
+    result =
+      if err.isNil:
+        extccomp.addLocalCompileOption(
+          c.config, s, toFullPathConsiderDirty(c.config, owner.fileIdx))
+        recordPragma(c, it, "localpassl", s)
+        it
+      else:
+        err
+  of wPush, wPop, wPragma:
+    # these require more context and thus special handling by the callsite
+    result = it
+  of wChecks, wObjChecks, wFieldChecks, wRangeChecks, wBoundChecks,
+      wOverflowChecks, wNilChecks, wAssertions, wWarnings, wHints,
+      wLineDir, wOptimization, wStaticBoundchecks, wStyleChecks,
+      wCallconv, wDebugger, wProfiler,
+      wFloatChecks, wNanChecks, wInfChecks, wTrMacros,
+      wStackTrace, wLineTrace:
+    var tmp = c.config.options
+    result = processOption(c, it, tmp)
+    c.config.options = tmp
+  of wEmit:
+    result = pragmaEmit(c, it)
+  of wLinearScanEnd, wComputedGoto:
+    result = noVal(c, it)
+  of wByRef:
+    # XXX: the special word is not in any statement-related allow list
+    result = noVal(c, it)
+    var tmp = c.config.options
+    result = processOption(c, it, tmp)
+    c.config.options = tmp
+  of wExperimental:
+    if not isTopLevel(c):
+      result = c.config.newError(it,
+                  PAstDiag(kind: adSemExperimentalRequiresToplevel))
+    result = processExperimental(c, it)
+  else:
```

Comment:
I was about to say that this switch should be exhaustive, but then I did an inventory of pragmas and here is what I got:

The constant `stmtPragmas` is missing the following pragmas from this case statement:
- `wCallConv` (same as pushing a calling convention, cooooool)
- `wDebugger` (dead with the embedded nim debugger, noop explicitly in the case?)
- `wProfiler`
- `wByRef`

The constant `stmtPragmas` has the following that are missing from this case statement:
- `wLinksys`
- `wPure` 
- `wDeadCodeElimUnused` (this one is deprecated per the comment)
- `wEffects` (pragma remains and `sempass2` sees it and emits effects at that point)
- `wAssert` (does nothing, might have been an old way to force an assertion)

The first set can be safely added to `stmtPragmas`. For the second set I would add `wEffects` to the case, I don't think it belongs in the symbol pragmas, the rest should be safe to drop from the constant. At which point this case can be exhaustive.

Reply:
I made the case statement exhaustive, although with two differences:
- `wDeadCodeElimUnused` is kept as a no-op, with a comment suggesting its removal
- `wAssert` is kept -- it can be used to enable/disable assertions for a section of code

Reply:
Oh, I misread the code, `wAssert` is indeed unused.